### PR TITLE
Rewrite NewServer to reduce cyclomatic complexity

### PR DESCRIPTION
**Pull request by pristupaanastasia**
#### Reference issue

https://github.com/pion/turn/issues/286

Fixes #...

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -172,6 +150,25 @@ func (s *Server) Close() error {
...
+	allocationManager, err := allocation.NewManager(allocation.ManagerConfig{
+		AllocatePacketConn: addrGenerator.AllocatePacketConn,
+		AllocateConn:       addrGenerator.AllocateConn,
+		PermissionHandler:  permissionHandler,
+		LeveledLogger:      s.log,
+	})
+	if err != nil {
+		return allocationManager, err
+	}
+	s.allocationManagers[i] = allocationManager
```
Comment (stv0g) [inline server.go:168]: I am just wondering what happens if the user provides both `listenerConfigs` and `packetConnConfigs`?
Arn't those indices then overwritten by each other? Or are those two settings mutually exclusive?

Comment (stv0g) [reply server.go:168]: Oh, nevermind. I figured it out..

File: server.go [hunk 2]
```diff
@@ -68,22 +68,11 @@ func NewServer(config ServerConfig) (*Server, error) {
 
 	for i := range s.packetConnConfigs {
 		go func(i int, p PacketConnConfig) {
```
Comment (stv0g) [inline server.go:70]: I think its a little nicer to create two new functions for the whole goroutine here.
E.g.  `readPacketConn` and `readListener`.

Those functions should also get rid of the `i` variable. Thats just not a nice way to handle the `allocationManagers`. I would just to a `s.allocationManagers = append(s.allocationManagers, ...)`.

**Comment by stv0g** [comment]
@pristupaanastasia I made some more changes to your PR and squashed it into a single commit.
It should be ready for merging now.

@rg0now Do you mind to have a look at the changes before we merge them?

**Comment by rg0now** [comment]
One thing that strikes me as odd is that we seem to call `createAllocationManager` twice for `ListenerConfigs`, once in NewServer (I guess that's the place where we'll eventually want to call it) [here](https://github.com/pristupaanastasia/turn/blob/012bea32ba4bc424713ba632fb4c10094adb71c8/server.go#L77), and then once again in `readListener` [here](https://github.com/pristupaanastasia/turn/blob/012bea32ba4bc424713ba632fb4c10094adb71c8/server.go#L136). This should append two allocation managers per `ListenerConfig`, which should not cause a problem I guess but it's still curious. 

Apart,  there must be good reason for renaming `errors` to `errs` throughout, I'd be interested to know why (maybe I'm doing something wrong in my code?).

Finally, I guess we can simplify [this](https://github.com/pristupaanastasia/turn/blob/012bea32ba4bc424713ba632fb4c10094adb71c8/server.go#L159) like below, is there a reason why we introduce a separate variable? (I really miss the ternary operator from Go but I guess it's only just me). 

```
if handler == nil {
	handler = DefaultPermissionHandler
}

am, err := allocation.NewManager(allocation.ManagerConfig{
	AllocatePacketConn: addrGenerator.AllocatePacketConn,
	AllocateConn:       addrGenerator.AllocateConn,
	PermissionHandler:  handler,
	LeveledLogger:      s.log,
})
```

Note that I may be wrong on each of these points, haven't done any tests myself.

**Comment by stv0g** [comment]
> One thing that strikes me as odd is that we seem to call createAllocationManager twice for ListenerConfigs, once in NewServer (I guess that's the place where we'll eventually want to call it) [here](https://github.com/pristupaanastasia/turn/blob/012bea32ba4bc424713ba632fb4c10094adb71c8/server.go#L77), and then once again in readListener [here](https://github.com/pristupaanastasia/turn/blob/012bea32ba4bc424713ba632fb4c10094adb71c8/server.go#L136).

Ouch :( Sorry this was a mistake by me. It should only be called once per listener in `NewServer`.
@pristupaanastasia initially called it within the Goroutine which required locking. I moved it outside, so we can safe the lock.

> Apart, there must be good reason for renaming errors to errs throughout, I'd be interested to know why (maybe I'm doing something wrong in my code?).

Agreed. I reverted the change.

> Finally, I guess we can simplify [this](https://github.com/pristupaanastasia/turn/blob/012bea32ba4bc424713ba632fb4c10094adb71c8/server.go#L159) like below, is there a reason why we introduce a separate variable? (I really miss the ternary operator from Go but I guess it's only just me).

Nice catch! No its, not just you. I miss it dearly :D

**Review by stv0g — APPROVED** [review]

**Pull request merged**